Add a web label

11EJDE11 opened the pull request:
This PR adds a label that can query JSON data from a datasource. We can use this for the online player count, for ladder ranking, or news. INI datasources are also supported.

Datasources are specified in `ClientDefinitions.ini`

```
[DataSources]
DataSourceID=URL,RefreshIntervalSeconds,TimeoutSeconds,Format ;example: TopPlayers=https://ladder.cncnet.org/api/v1/qm/ladder/rankings,300,10,json|ini
```


Labels then point to a datasource and can query the JSON.

```
[ExtraControls]
1=lblTopPlayer1:XNAClientJSONLabel

[lblTopPlayer1]
$Type=XNAClientJSONLabel
DataSourceID=TopPlayers
Template=1. {YR[0].player_name} - {YR[0].points:N0} pts
LoadingText=Loading...
FallbackText=---
FontIndex=1
RemapColor=230,230,230
Location=360,166
```

See `INISytem.md` for more info.


Excuse the crappy screenshot.
<img width="223" height="139" alt="image" src="https://github.com/user-attachments/assets/dbe6ac16-b490-4f13-8448-2e073a097abe" />

11EJDE11 committed:
Add XNAClientJSONLabel

11EJDE11 committed:
Extras

Metadorius commented on ClientGUI/XNAClientJSONLabel.cs:
```diff
@@ -0,0 +1,191 @@
+            Logger.Log($"JSONPath error: {ex.Message}");
+            return null;
+        }
+    }
```
I think this should happen way earlier at manager stage, and the control should only receive a parsed json and do template substitution, otherwise we're duplicating work.

Also we already use System.Text.Json, I see that it has no support for JSON Path for some reason, maybe this extension to System.Text.Json would do better? https://www.nuget.org/packages/JsonPath.Net Just so we don't introduce two similar libraries.

11EJDE11 replied:
JSON is now parsed to JToken in the manager and the label only does template substitution, so the duplication is sorted. Regarding JsonPath.NET, I couldn't get this working in an acceptable manner.

Metadorius commented on ClientGUI/XNAClientJSONLabel.cs:
```diff
@@ -0,0 +1,191 @@
+    public string Template { get; set; }
+    public string LoadingText { get; set; }
+    public int MaxResults { get; set; } = 0;
+    public string FallbackText { get; set; } = "N/A";
```
missing .L10N call

Metadorius commented on ClientCore/JSONDataSourceManager.cs:
```diff
@@ -0,0 +1,339 @@
+        private string ConvertIniToJson(string iniContent)
+        {
+            var iniFile = new IniFile();
+            using (var reader = new StringReader(iniContent))
```
There is a constructor in IniFile that takes a stream, no need to parse it by hand, just take a string and make a stream out of it 😄

11EJDE11 replied:
[65229d1](https://github.com/CnCNet/xna-cncnet-client/pull/883/commits/65229d1294d1a9ddc4a1cfd53be6ab2e3fe8583f)

Metadorius commented on ClientCore/JSONDataSourceManager.cs:
```diff
@@ -0,0 +1,339 @@
+
+            var sb = new StringBuilder();
+            sb.Append("{");
+            bool firstSection = true;
```
The aforementioned bringing of JSON parsing earlier than that will allow you to construct the programmatic representation via a few foreaches, iterating sections and then keys to set values and greatly simplifying the code 🙂

11EJDE11 replied:
[65229d1](https://github.com/CnCNet/xna-cncnet-client/pull/883/commits/65229d1294d1a9ddc4a1cfd53be6ab2e3fe8583f)

Metadorius commented on ClientCore/JSONDataSourceManager.cs:
```diff
@@ -0,0 +1,339 @@
+            if (string.IsNullOrEmpty(value))
+                continue;
+
+            // format: URL,RefreshIntervalSeconds,TimeoutSeconds,Format
```
URL could have commas in it, I think instead should do something like

```ini
[DataSourceName]
URL=https://...
RefreshIntervalSeconds=
```

Also I think individual data source parsing should be in the DataSource itself?

11EJDE11 replied:
[65229d1](https://github.com/CnCNet/xna-cncnet-client/pull/883/commits/65229d1294d1a9ddc4a1cfd53be6ab2e3fe8583f)

Metadorius commented on ClientCore/JSONDataSourceManager.cs:
```diff
@@ -0,0 +1,339 @@
+public class JSONDataSourceManager
+{
+    private static JSONDataSourceManager _instance;
+    public static JSONDataSourceManager Instance => _instance ??= new JSONDataSourceManager();
```
We have a DI container in use, I think you could use that? The codebase haven't fully migrated though, so tell me if you encounter any issues with that

11EJDE11 replied:
Looks good, works well. [a19f21c](https://github.com/CnCNet/xna-cncnet-client/pull/883/commits/a19f21c2cc42ff42d123962573f16137c79dedb6)

Metadorius commented on ClientCore/JSONDataSourceManager.cs:
```diff
@@ -0,0 +1,339 @@
+        private readonly string _url;
+        private readonly int _refreshIntervalSeconds;
+        private readonly int _timeoutSeconds;
+        private readonly string _format;
```
I think this `_format` should be either an enum and/or maybe even a tiny simple interface/class with one method that converts input value (stream?) and gives JSON for arbitrary data representation.

SadPencil replied:
I agree with an interface. So we can decouple some string parsing logic from the UI logic. [My recent editing experience on LAN lobby](https://github.com/CnCNet/xna-cncnet-client/pull/826#issuecomment-3758744082) shows that it might waste a huge amount of time to maintain codes if different kinds of logics are not decoupled

11EJDE11 replied:
Done in the latest commit - IDataSourceFormat.

Metadorius commented on ClientCore/JSONDataSourceManager.cs:
```diff
@@ -0,0 +1,339 @@
+    {
+        private readonly string _id;
+        private readonly string _url;
+        private readonly int _refreshIntervalSeconds;
```
does it run even in background or when controls using it is not active?

although I realise now that it will probably complicate the code a lot, so I don't think there's a big need in doing it exactly. but if there is *a lot* of players that don't go online -- maybe autorefresh isn't something we may want to do? it will strain our servers quite a bit, maybe we should update just when the control is shown for the first time? (or sth in this direction)

Metadorius commented on ClientGUI/XNAClientJSONLabel.cs:
```diff
@@ -0,0 +1,191 @@
+
+namespace ClientGUI;
+
+public class XNAClientJSONLabel : XNALabel
```
Perhaps let's name it `XNAClientWebLabel`, because it's not just JSON anymore?

Metadorius replied:
also maybe it could be good to have it as a descendant of XNALinkLabel? in case the users want to see details

11EJDE11 replied:
Renamed [65229d1](https://github.com/CnCNet/xna-cncnet-client/pull/883/commits/65229d1294d1a9ddc4a1cfd53be6ab2e3fe8583f)
Will look into the LinkLabel.

11EJDE11 replied:
What do you mean by "in case the users want to see details"? @Metadorius

Metadorius replied:
I meant that, for example, there's ladder info like you've shown, and clicking on that info could open the full ladder info with details in browser, which is the details in question.

11EJDE11 replied:
Makes sense, sorted thanks.

Metadorius reviewed (changes requested):
Nice! Gave you some feedback.

Generally though while this approach definitely has it's place, I am not sure if we should do an explicit separate data source manager? My thoughts were more about something like an implicit manager class; different labels could have content URL endpoints specified right in place, and the response will be shared between them by that class (say, if the requests came to the manager in span of half a minute, then give the same data to every subscriber; if something updates the URL -- then all of the consumers of that URL will receive that). 

Not fully set on the above, what are your thoughts?

cc @SadPencil @MahBoiDeveloper @Starkku @Rampastring

11EJDE11 committed:
Merge branch 'develop' into feat-web-label-2

11EJDE11 committed:
Merge remote-tracking branch 'origin/develop' into feat-web-label-2

11EJDE11 committed:
Switch to RS for INI files
Fix crash when no datasources specified

11EJDE11 committed:
Fixup

SadPencil commented on ClientGUI/XNAClientJSONLabel.cs:
```diff
@@ -0,0 +1,191 @@
+    public string Template { get; set; }
+    public string LoadingText { get; set; }
+    public int MaxResults { get; set; } = 0;
+    public string FallbackText { get; set; } = "N/A";
```
"N/A" is short for not applicable, instead of "not available". So I don't think we should use "N/A" here

Metadorius replied:
it's used interchangeably as either of those, so I think it's fine?

pzhlkj6612 replied:
The example in "Docs/INISystem.md" uses `---`. I think using `---` as the default value is better because it exactly means "nothing" with placeholders. Then we can change the example value to `N/A`.

11EJDE11 replied:
Changed to "---" [65229d1](https://github.com/CnCNet/xna-cncnet-client/pull/883/commits/65229d1294d1a9ddc4a1cfd53be6ab2e3fe8583f)

11EJDE11 committed:
Merge develop into feat-web-label-2

11EJDE11 committed:
Update docs
Update datasources format in ClientDefintions
Add localisation
Change to XNAClientWebLabel

11EJDE11 committed:
Move to DI

11EJDE11 committed:
Remove dead Template, LoadingText, FallbackText as TranslationINIParser already handles them.
Replace _format string with IDataSourceFormat interface.
IniDataSourceFormat now builds a JObject directly

11EJDE11 committed:
Inherit from XNAClientLinkLabel instead of XNALabel

11EJDE11 committed:
Update docs
